CRM/Member/Form/Membership.php

```diff
@@ -182,6 +222,9 @@ public function preProcess() {
     // This string makes up part of the class names, differentiating them (not sure why) from the membership fields.
     $this->assign('formClass', 'membership');
     parent::preProcess();
+    if ($this->isUpdateToExistingRecurringMembership()) {
+      $this->entityFields['end_date']['is_freeze'] = TRUE;
```

Comment:
We can't do this freeze until preProcess as id is not retrieved from the form earlier. I did go down the path of trying to move the call to setEntityFields to the preProcess - but perhaps it makes sense to keep it generic & tweak it in preProcess - as done here.

There are other fields that will ideally be moved into this & into the setEntityFields